contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,549 @@
+      ENVOY_LOG(debug, "tcp upstream has been destroyed");
+      return;
+    }
+    filter_->has_destroyed_ = true;
+  }                
+
+  auto reason = (filter_->decoding_state_.isProcessingInGo() || filter_->encoding_state_.isProcessingInGo())
+                ? DestroyReason::Terminate
+                : DestroyReason::Normal;
+  dynamic_lib_->envoyGoOnTcpUpstreamDestroy(filter_->req_, int(reason));
+
+}
+
+Envoy::Http::Status TcpUpstream::encodeHeaders(const Envoy::Http::RequestHeaderMap& headers,
+                                               bool end_stream) {
+  ENVOY_LOG(debug, "tcp upstream encodeHeaders, header size: {}, end_stream: {}", headers.size(), end_stream);
+
+  filter_->initRequest();
+
+  ProcessorState& state = filter_->encoding_state_;
+  state.headers = &headers;
+  Buffer::OwnedImpl buf;
+  Buffer::Instance& buffer = state.doDataList.push(buf);
+  auto s = dynamic_cast<processState*>(&state);
+  state.setFilterState(FilterState::ProcessingHeader);
+
+  GoUint64 go_tatus = dynamic_lib_->envoyGoEncodeHeader(
+    s, end_stream ? 1 : 0, headers.size(), headers.byteSize(), reinterpret_cast<uint64_t>(&buffer), buffer.length());
+  state.setFilterState(FilterState::Done);
+  state.doDataList.moveOut(buf);
+
+  bool upstream_conn_half_close;
+  switch (go_tatus) {
+  case static_cast<GoUint64>(SendDataStatus::SendDataWithTunneling):
+    // send data to upstream with conn not half close
+    upstream_conn_half_close = false;
+    upstream_conn_data_->connection().write(buf, upstream_conn_half_close);
+    break;
+
+  case static_cast<GoUint64>(SendDataStatus::SendDataWithNotTunneling):
+    // send data to upstream with conn half close
+    upstream_conn_half_close = true;
+    upstream_conn_data_->connection().write(buf, upstream_conn_half_close);
+    break;
+
+  case static_cast<GoUint64>(SendDataStatus::NotSendData):
+    // not send data to upstream
+    break;
+
+  default:
+    // send data to upstream
+    ENVOY_LOG(error, "encodeData unexpected go_tatus: {}", go_tatus);
+    upstream_conn_data_->connection().write(buf, end_stream);
+    break;
+  }
+
+
+  // Headers should only happen once, so use this opportunity to add the proxy header, if configured.
+  ASSERT(route_entry_ != nullptr);
+  if (route_entry_->connectConfig().has_value()) {
+    Buffer::OwnedImpl data;
+    const auto& connect_config = route_entry_->connectConfig();
+    if (connect_config->has_proxy_protocol_config() &&
+        upstream_request_->connection().has_value()) {
+      Extensions::Common::ProxyProtocol::generateProxyProtoHeader(
+          connect_config->proxy_protocol_config(), *upstream_request_->connection(), data);
+    }
+
+    if (data.length() != 0 || end_stream) {
+      ENVOY_LOG(error, "encodeData send proxy : {}", data.toString());
+      upstream_conn_data_->connection().write(data, upstream_conn_half_close);
+    }
+  }
+
+  if (buf.length() != 0) {
+    upstream_conn_data_->connection().write(buf, false);
```

Comment:
false should be end_stream? since it might be a Header only request?

Reply:
done